src/main/java/com/iridium/iridiumskyblock/listeners/BlockFormListener.java

```diff
@@ -1,42 +1,78 @@
+            for (Map.Entry<XMaterial, Integer> entries : upgrades.getValue().ores.entrySet()) {
+                Material material = entries.getKey().parseMaterial();
+                for (int i = 0; i < entries.getValue(); i++) {
+                    oresArrays[0][index++] = material;
+                }
+            }
+            oresArrays[1] = new Material[netherOreFrequencySum];
+            int indexNether = 0;
+            for (Map.Entry<XMaterial, Integer> entries : upgrades.getValue().netherOres.entrySet()) {
+                Material material = entries.getKey().parseMaterial();
+                for (int i = 0; i < entries.getValue(); i++) {
+                    oresArrays[1][indexNether++] = material;
+                }
+            }
+            upgradesParsed.put(upgrades.getKey(), oresArrays);
+        }
+    }
+
     @EventHandler
     public void onBlockForm(BlockFormEvent event) {
         XMaterial newMaterial = XMaterial.matchXMaterial(event.getNewState().getType());
         // Custom basalt generators should only work in nether
-        if (newMaterial.equals(XMaterial.COBBLESTONE) || (newMaterial.equals(XMaterial.BASALT) && event.getBlock().getLocation().getWorld().getEnvironment().equals(World.Environment.NETHER))) {
+        if (newMaterial.equals(XMaterial.COBBLESTONE) || newMaterial.equals(XMaterial.STONE) || (newMaterial.equals(XMaterial.BASALT) && event.getBlock().getLocation().getWorld().getEnvironment().equals(World.Environment.NETHER))) {
             Optional<Island> island = IridiumSkyblock.getInstance().getIslandManager().getIslandViaLocation(event.getNewState().getLocation());
             if (island.isPresent()) {
-                IslandUpgrade islandUpgrade = IridiumSkyblock.getInstance().getIslandManager().getIslandUpgrade(island.get(), "generator");
-                OresUpgrade oresUpgrade = IridiumSkyblock.getInstance().getUpgrades().oresUpgrade.upgrades.get(islandUpgrade.getLevel());
-                Map<XMaterial, Integer> ores = newMaterial.equals(XMaterial.BASALT) ? oresUpgrade.netherOres : oresUpgrade.ores;
-                List<XMaterial> materials = new ArrayList<>();
-                Random random = new Random();
-
-                for (Map.Entry<XMaterial, Integer> entries : ores.entrySet()) {
-                    for (int i = 0; i < entries.getValue(); i++) {
-                        materials.add(entries.getKey());
-                    }
-                }
-
-                Material material = materials.get(random.nextInt(materials.size())).parseMaterial();
+                Material[] materials = upgradesParsed.get(IridiumSkyblock.getInstance().getIslandManager().getIslandUpgrade(island.get(), "generator").getLevel())[newMaterial.equals(XMaterial.BASALT) ? 1 : 0];
+                Material material = materials[random.nextInt(materials.length)];
+                if (material == Material.COBBLESTONE && event.getBlock().getType() == Material.STONE)
+                    material = Material.STONE;
```

Comment:
No more cobblestone? I don't play Skyblock myself, is that normal? 👀

Reply:
Took me a while to realise what this was doing aswell, there should probabbly be a comment or something xD

What its doing is if there is a stone generator, and the new material is cobblestone, instead of setting it to cobblestone it should set it to stone instead since it was actually a stone generator normal cobblegens wont be affected

Reply:
Oh, alright, my bad